src/lang-promql/lsp/client.ts

```diff
@@ -0,0 +1,61 @@
+import axios from 'axios';
+import { AutocompleteContext, Completion, CompletionResult } from "@codemirror/next/autocomplete";
+import { AutocompleteResponse, TextEdit } from "./model";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isIterable(obj: any): boolean {
+  return obj != null && typeof obj[ Symbol.iterator ] === 'function';
+}
+
+export class LSPClient {
+  private readonly url: string
+  private readonly autocompleteEndpoint = "/completion"
+
+  constructor(url: string) {
+    this.url = url;
+  }
+
+  autocomplete(context: AutocompleteContext): Promise<CompletionResult> {
+    const {state, pos} = context
+    const line = state.doc.lineAt(pos)
+    const body = {
+      expr: state.doc.sliceString(0),
+      limit: 100,
+      // positionChar start at 0 at the beginning of the line for LSP
+      positionChar: pos - line.from - 1 > 0 ? pos - line.from - 1 : pos - line.from,
+      // positionLine start at 0 from LSP and at 1 from CMN
+      positionLine: line.number - 1,
+    }
+    return axios.post<AutocompleteResponse[]>(this.url + this.autocompleteEndpoint, body)
+      .then((response) => {
+        const options: Completion[] = []
+        // for every textEdit present, they qll have exactly the same value
+        // so the goal is to save the last one present in order to use it later to calculate the position "from"
+        // Note: textEdit can be undefined but that's just because otherwise it doesn't compile
+        // see more here: https://github.com/microsoft/TypeScript/issues/12916
+        // and here https://github.com/microsoft/TypeScript/issues/9568
+        let textEdit: TextEdit | undefined
+        if (isIterable(response.data)) {
+          for (const res of response.data) {
+            // TODO map kind with icon Completion.type when it is released on CMN side
+            // https://github.com/codemirror/codemirror.next/commit/459bba29c1fd1d80fc4f36dac27e5825b2362273
+            let label = res.label
+            if (res.textEdit) {
+              label = res.textEdit.newText
+              textEdit = res.textEdit
```

Comment:
The label and the newText aren't always equal. In this example, they aren't.

```json
[
  {
    "label": "label_replace",
    "kind": 3,
    "sortText": "__1__000016426",
    "insertTextFormat": 2,
    "textEdit": {
      "range": {
        "start": {
          "line": 0,
          "character": 0
        },
        "end": {
          "line": 0,
          "character": 9
        }
      },
      "newText": "label_replace($1)"
    },
    "command": {
      "title": "",
      "command": "editor.action.triggerParameterHints"
    }
  }
]
```

Reply:
mmm yeah that's the whole point of this code. newText is more accurate than the label. So if exist, then we will always prefer it.
So far I don't see where is the issue.

Reply:
The `label` is what the editor should show in the completion dropdown.

The `newText` describes, what should be inserted, using a template format specified by the LSP.

Showing the `newText` as a `label` violates the LSP.

Reply:
ok, I managed to do it using the field `apply` from the `Completion`.
Just keep in mind, what we are doing here is just an approximation of the LSP implementation. It can happen we won't do it like LSP is proposing to do ;).

How the display and the feeling with the autocompletion will be, is more important than respecting what LSP is exposing. (at least for me)

Reply:
Anyway thanks for the explanations, it was interesting to know the different between the two fields.

Reply:
This still doesn't handle the cursor position hints correctly.

It seems to be possible to use a function in the `apply` field which should make it possible to correctly place the cursor after inserting a completion.

Reply:
As long as we're not parsing TextEdits correctly, it's still better to use the `label` field instead of the `newText` for inserting.

That's how Vim handles it, too.

Reply:
I prefer to use the newText for the moment, because it provides the bracket automatically and so it makes a good difference immediately between a function and a simple string like a label or something equivalent.
Let's see later how to change the cursor position. If we cannot change it, then at least the `$1` will be removed manually.

Reply:
Maybe it's best to leave it as is for now and open an issue to eventually do it properly.

Reply:
yeah. Here I'm more let's say in a "die and retry" mode. Meaning that I will try many things and then see if the feeling is nice or not.

You can expect that the code is going to change a lot

Reply:
in the meantime, code is opened and the PR as well. Feel free to fix it in case you would like to go furthermore and to enter in the nice world of CodeMirror !

Reply:
>  Feel free to fix it in case you would like to go furthermore and to enter in the nice world of CodeMirror !

That'd mean I'd have to actually learn JavaScript ;)

In case I'm be working on this, I think I'll focus more on the Codemirror upstream for now, since two people making large changes to the same code in parallel is likely to get messy.